llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -200,93 +171,142 @@ groupEntryPointsByKernelType(ModuleDesc &MD,
-// edges are "calls" relation.
-class CallGraph {
+// Represents "dependency" or "use" graph of global objects (functions and
+// global variables) in a module. It is used during device code split to
+// understand which global variables and functions (other than entry points)
+// should be included into a split module.
+//
+// Nodes of the graph represent LLVM's GlobalObjects, edges "A" -> "B" represent
+// the fact that if "A" is include into a module, then "B" should be included
+// as well.
+//
+// Examples of dependencies which are represented in this graph:
+// - Function FA calls function FB
+// - Function FA uses global variable GA
+// - Global variable GA references (intiialized with) function FB
+// - Function FA stores address of a function FB somewhere
+// - Function FA uses global variable GB
+//
+// The following cases are treated as dependencies between global objects:
+// 1. Global object A is used within by a global object B in any way (store,
+//    bitcast, phi node, call, etc.): "A" -> "B" edge will be added to the
+//    graph;
+// 2. function A performs an indirect call of a function with signature S and
+//    there is a function B with signature S. "A" -> "B" edge will be added to
+//    the graph;
+class DependencyGraph {
 public:
-  using FunctionSet = SmallPtrSet<const Function *, 16>;
-
-private:
-  std::unordered_map<const Function *, FunctionSet> Graph;
-  SmallPtrSet<const Function *, 1> EmptySet;
-  FunctionSet AddrTakenFunctions;
+  using GlobalSet = SmallPtrSet<const GlobalValue *, 16>;
+
+  DependencyGraph(const Module &M) {
+    // Group functions by their signature to handle case (2)
+    DenseMap<const FunctionType *, DependencyGraph::GlobalSet>
+        FuncTypeToFuncMap;
+    for (const auto &F : M.functions()) {
+      FuncTypeToFuncMap[F.getFunctionType()].insert(&F);
+    }
 
-public:
-  CallGraph(const Module &M) {
-    for (const auto &F : M) {
-      for (const Value *U : F.users()) {
-        if (const auto *I = dyn_cast<CallInst>(U)) {
-          if (I->getCalledFunction() == &F) {
-            const Function *F1 = I->getFunction();
-            Graph[F1].insert(&F);
-          }
-        }
-      }
-      if (F.hasAddressTaken()) {
-        AddrTakenFunctions.insert(&F);
+    // We add every function into the graph
+    for (const auto &F : M.functions()) {
+      // case (1)
+      for (const Value *U : F.users())
+        addUserToGraphRecursively(cast<const User>(U), &F);
+
+      // case (2)
+      for (const auto &I : instructions(F)) {
+        if (!isa<CallInst>(&I))
+          continue;
+
+        const auto *CI = cast<CallInst>(&I);
+        // Direct calls were handled above
+        if (!CI->isIndirectCall())
+          continue;
+
+        const FunctionType *Signature = CI->getFunctionType();
+        const auto &PotentialCallees = FuncTypeToFuncMap[Signature];
+        Graph[&F].insert(PotentialCallees.begin(), PotentialCallees.end());
```

Comment:
i dont know if this is a real issue or not, but you can imagine if you have a signature with a ton of callees, we end up having multiple copies of that set, which may increase memory usage. i wonder if we can improve that, and if we could, if its worth it.

Reply:
At the moment both function pointers and virtual functions are not publicly-available features, i.e. there are no formal specifications for them, flags to enable them are hidden under `-cc1` mode (require `-Xclang` to pass them), etc.

The intent here is to document this signature-matching approach in corresponding language extension specifications so that users are aware about how compiler behaves and can adjust their code accordingly. For example, they can introduce unused _tag_ arguments do make indirectly called functions unique in terms of a signature to avoid unnecessary copies during device code split.

_Theoretically_, we could have improved our analysis further to actually track which exact functions are address taken and only consider them as candidates for indirect calls. It _seems_ doable? But I'm concerned that I could be missing some corner case, which is harder to trace, let me think about it more

Reply:
yeah i dont think the case i was thinking of is likely to happen, just something to consider since we've seen multiple sycl-post-link performance issues

Reply:
Added `TODO` to reduce list of potential callees in 3c7a4d5d396c117180e30689319d52019c466abc and removed kernels from that list in df0b4e3643804398d234b947012fd3e62eefa444